### PR TITLE
Changes for running Zephyr as Xen control domain

### PR DESCRIPTION
**Pull request by firscity**
Currently, Zephyr RTOS can be run as Domain-U (unprivileged domain) under Xen hypervisor via xenvm board. But Zephyr looks like great safe Linux alternative for Xen Domain-0 (privileged domain).

Zephyr kernel now includes lot of required drivers (Xen events, grant tables etc.), but start as a control domain (Domain-0) requires additional changes: domain control interface, possibility to configure any Zephyr board to run as Xen control domain etc. They were added in this PR.

Patch `arch: arm64: mmu: WA: avoid using of set/way cache instructions for Xen` is only for demonstration/testing purposes and will be replaced with proper fix. Related discussion about set/way problem under Xen hypervisor: #58436.

### CODE REVIEW

File: drivers/CMakeLists.txt
Comment (carlocaione) [inline drivers/CMakeLists.txt:15]: uhm, for consistency this should be `add_subdirectory_ifdef(CONFIG_XEN)` (or similar) and then use `CONFIG_DT_HAS_XEN_XEN_ENABLED` in the actual driver (see how this is done for the other drivers).

**Review by carlocaione — CHANGES_REQUESTED** [review]
This is a bit out of my comfort zone, so just some trivial nits for now.

**Review by povergoing — COMMENTED** [review]
Zephyr as DOM0 is really impressive, only some nits I comment inline

**Comment by povergoing** [comment]
A quick question, does this DOM0 support SMP mode?

File: include/zephyr/xen/public/arch-arm.h
Comment (firscity) [inline include/zephyr/xen/public/arch-arm.h:299]: sdf

**Comment by firscity** [comment]
First of all, @carlocaione @povergoing thank you for review and fresh look on these changes.

To prevent mail spam (can't reply once with "Start review" on "outdated" comments), here are a quoted replies:


> What about instead moving this somewhere in arch/arm64/core? Having this in drivers is odd.

>no need to have default n.

> uhm, for consistency this should be add_subdirectory_ifdef(CONFIG_XEN) (or similar) and then use CONFIG_DT_HAS_XEN_XEN_ENABLED in the actual driver (see how this is done for the other drivers).

> Are the depends on SHIELD_XEN_DOM0 and depends on XEN_INITIAL_DOMAIN somehow duplicated?
Any case is there that XEN_INITIAL_DOMAIN is true but SHIELD_XEN_DOM0 is false?

> Do INITIAL_DOMAIN and DOM0 mean the same concept? If they are the same, sometimes INITIAL_DOMAIN and sometimes DOM0 are a bit weird.
...

These comments are completely right, we needed a fresh look on what is already done. I refactored this part and removed all redundant configs with just `CONFIG_XEN` and `CONFIG_XEN_DOM0`. It's enough to determine a type of Xen setup and can be easily selected when "xen,xen" node is present (`XEN_DOM0` - when node is present and shield is applied). Thank you, fixed.


> this looks weirdly indented (like the rest of the file).

> Keep spacing consistent.

> BIT() ?

This file is just a copy of original Xen public header. I fixed coding style, but left bit values definitions as is (at some time it will probably require an update from fresh Xen version).


> cortex-a57?

Oh, sorry.  Sure, `cortex-a53`, fixed.


> A quick question, does this DOM0 support SMP mode?

We hadn't need to enable more than one core for both xenvm and Xen Dom0 setups, so I don't have an answer for that, needs testing.

File: arch/arm64/core/Kconfig
Comment (carlocaione) [inline arch/arm64/core/Kconfig:168]: Oh man, this is super weird to have here.

Ok, it's time to have XEN as first class citizen then. Can you move all the XEN related stuff into a `xen/Kconfig.xen` file that you `rsource` from here? At least we do not pollute the general `Kconfig` too much.

**Review by carlocaione — CHANGES_REQUESTED** [review]

File: arch/arm64/core/Kconfig
Comment (firscity) [reply arch/arm64/core/Kconfig:168]: @carlocaione had such ideas too during refactor. I will fix it soon along with CI. Thank you.

File: drivers/xen/events.c
Comment (luca-fancellu) [inline drivers/xen/events.c:50]: Please correct me if I am wrong, but can this be protected with the dom0 Kconfig? Would it make sense?

File: drivers/xen/memory.c
Comment (luca-fancellu) [inline drivers/xen/memory.c:5]: NIT: can a blank line be added before the include?

**Review by luca-fancellu — COMMENTED** [review]
Nice work!

**Comment by jgrall** [comment]
Commit message for "drivers: xen: gnttab: do Xen node mapping inside driver"

> Unfortunately, after moving we also need to increase number of XLAT
tables. The previous code was more efficient if we talking about
usage of XLAT tables, because it mapped grant tables using a higher-
order table that allows mapping blocks of 2MB.

From the commit message it is not entirely clear why you couldn't continue to use 2MB afterwards. That said, the original code was buggy because AFAIK we don't guarantee that the grant-table area will be 2MB aligned. So you may have mapped memory with the wrong attributes.

File: arch/arm64/core/mmu.c
Comment (jgrall) [inline arch/arm64/core/mmu.c:819]: So there are a few issues with using set/way instructions in Zephyr:
  * As you said above they are not easily virtualizable
  * On Armv8, sytem caches are can only be cleaned by VA. So this would not flush all the caches.

If you want more details, there is a pretty good talk from Marc Zyngier about it:
https://www.youtube.com/watch?v=A_zCxzpxzmE

Therefore, I think whatever solution you come up with should be generic.

Comment (firscity) [reply arch/arm64/core/mmu.c:819]: Thank you for addition, we currently have discussion #58436 for this issue. We tried U-boot approach, but it will have issues during dynamic memory mapping. For sure, final solution should be generic, this commit is here only for test purposes during review.

**Comment by firscity** [comment]
@jgrall thank you for your review!

> What's your plan for other Xen versions and going forward?

Now we want to add support for 4.17 release, based on its fixed domctl interface. Starting from 4.18 it is planned to update it only for stable releases and separate changes basing on major/minor Xen version defined during build.

> From the commit message it is not entirely clear why you couldn't continue to use 2MB afterwards. That said, the original code was buggy because AFAIK we don't guarantee that the grant-table area will be 2MB aligned. So you may have mapped some memory with the wrong atributes (for instance if the rest of the 2MB is MMIO).

Yes, sure, regions may be unaligned to 2MB. Both previous and current approaches are correct, but works differently. Previously, Zephyr used 1:1 mapping and in `xenvm` case it was possible to use L2 because address was 2MB aligned. Now we are using `device_map()` which picks virtual region from internal allocator (`virt_region_alloc()`) and usually returned regions are 4K page aligned. It leads to mapping grant-table area with L3 entries and consuming more XLAT tables.
Unfortunately, Zephyr `device_map()` does not offer interface for specifying virt_region alignment during mapping.

**Comment by jgrall** [comment]
> Now we want to add support for 4.17 release, based on its fixed domctl interface. Starting from 4.18 it is planned to update it only for stable releases and separate changes basing on major/minor Xen version defined during build.

I am a bit concerned of the complexity going forward. In the description, you suggested the goal is to provide a toolstack on top of Zephyr. Can you clarify your end goal? Is it creating domain based on a statically defined configuration?

**Comment by firscity** [comment]
@jgrall 

> You will most likely need to provide wrapper that will abstract the difference in version from the callers.

I think it can be done in Zephyr lib level and lib should have relatively stable interface for user application and be synced with Zephyr kernel.

> Before adding the helpers in Zephyr, I think it would be good to clarify your end goal. You are saying you will implement a toolstack. But how powerful will it be? E.g. will it create domain dynamically or will they be based on a statically defined configuration (like dom0less)?

Currently we are working on wrapper library, that can be included in user application and allows to manage domains, start console and xenstore server, work with device tree etc. It can load configuration, domain kernels and device trees dynamically (e.g. from filesystem) or create domain with pre-defined configuration.
It's called `zephyr-xenlib` and located [here](https://github.com/xen-troops/zephyr-xenlib).

File: arch/arm64/core/mmu.c
Comment (xakep-amatop) [reply arch/arm64/core/mmu.c:819]: > Therefore, I think whatever solution you come up with should be generic.

Here is the proper fix instead of the current workaround: https://github.com/zephyrproject-rtos/zephyr/pull/61035

File: boards/shields/xen_dom0/doc/index.rst
Comment (jfischer-no) [inline boards/shields/xen_dom0/doc/index.rst:9]: The community does not like "virtual" shields, instead snippets should be used, even if they do not offer the same functionality. https://docs.zephyrproject.org/latest/build/snippets/index.html

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: boards/shields/xen_dom0/doc/index.rst
Comment (xakep-amatop) [reply boards/shields/xen_dom0/doc/index.rst:9]: Just for clarification, when we were developing this functionality, the `snippets` concept didn't exist.
Thank you for pointing that out; we'll investigate this approach.

**Comment by firscity** [comment]
@xakep-amatop provided proper fix for set/way issues in #61035, so this PR is blocked until set/way patches will be merged.

I will drop WA set/way patch and will try to rework Dom0 build configuration to snippets (thanks @jfischer-no for suggestion), after first look I see that it is possible.

File: snippets/xen_dom0/snippet.yml
Comment (xakep-amatop) [inline snippets/xen_dom0/snippet.yml:7]: you can add `rcar_h3ulcb_ca57` and `salvator_xs_m3` too, support of them have been added to mainline recently

**Review by carlocaione — DISMISSED** [review]
I mean, in general LGTM but again I'm way out my conform zone here. So giving my +1 but I think we need a XEN maintainer in Zephyr.

**Review by xakep-amatop — APPROVED** [review]

File: snippets/xen_dom0/snippet.yml
Comment (firscity) [reply snippets/xen_dom0/snippet.yml:7]: @xakep-amatop, added in separate commit.

**Comment by firscity** [comment]
@carlocaione, @xakep-amatop thank you for approval. Actually me and other my team member are Xen maintainers in Zephyr. These changes passed internal review.

But I will be very grateful if @jgrall will take a look at this and #61035 (blocks this PR), he is one of Xen maintainers for ARM64.

**Review by povergoing — DISMISSED** [review]
LGTM

**Comment by firscity** [comment]
Dear all, since #61035 was merged today, this PR is unblocked and can be merged (requested changes with snippets were implemented).

**Comment by firscity** [comment]
@jfischer-no can you please re-check snippets implementation due to requested changes?

**Review by jfischer-no — DISMISSED** [review]
LGTM, but maybe @mbolivar-ampere has time to take a look at the new snippet.

**Comment by firscity** [comment]
Changes were rebased on top of `main` branch due to conflicts in testcase.yml file. No functional changes were made.

**Comment by firscity** [comment]
In last force-push: fixed issues in `kernel.device` CI tests (add exclude for newly added due to same reason of invalid build configuration).

As previous approvals were dismissed, can @carlocaione @povergoing @jfischer-no check it again (no functional changes were made, only rebase)? Or it is ready to merge?

**Comment by carlocaione** [comment]
@firscity can you fix the `WARNING: Title underline too short.` warning?

**Comment by firscity** [comment]
@carlocaione oh, I'm sorry, missed this.
Fixed

**Review by mbolivar-ampere — COMMENTED** [review]
@jfischer-no thanks for the ping. The snippet seems fine to me. At some point if this hits scalability limits we should consider adding the board directory to SNIPPET_ROOT automatically, to allow board support to be factored out from the main snippet implementation. But for now this seems fine. Let's see how it scales in practice.

**Comment by carlocaione** [comment]
@firscity I do not understand why the documentation step is failing. Could you please try to rebase on the latest `main`?

**Comment by carlocaione** [comment]
> @carlocaione, I already did this, but it did not helped, rebased once again. As I said previously, documentation builder incorrectly parses C struct in Xen header, which leads to a warning. And warning stops documentation build.

Ooh, ok, then this will need to be merged with superpowers.

**Review by carlocaione — APPROVED** [review]

**Review by povergoing — APPROVED** [review]
refresh +1

**Comment by firscity** [comment]
@carlocaione can you, please, ask someone with superpower to merge this? All other CI test finished successfully.

**Comment by carlocaione** [comment]
> @carlocaione can you, please, ask someone with superpower to merge this? All other CI test finished successfully.

@carlescufi @fabiobaltieri ^

**Pull request merged**
